File: src/main/scala/com/yogpc/qp/tile/TileExpPump.scala
package com.yogpc.qp.tile

import java.util

import com.yogpc.qp._
import com.yogpc.qp.block.{ADismCBlock, BlockExpPump, BlockPump}
import com.yogpc.qp.compat.InvUtils
import com.yogpc.qp.gui.TranslationKeys
import com.yogpc.qp.packet.PacketHandler
import com.yogpc.qp.packet.exppump.ExpPumpMessage
import javax.annotation.Nullable
import net.minecraft.entity.item.EntityXPOrb
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.EnumFacing
import net.minecraft.util.math.BlockPos
import net.minecraft.util.text.TextComponentString
import net.minecraft.world.World

import scala.collection.JavaConverters._

class TileExpPump extends APacketTile with IEnchantableTile with IDebugSender with IAttachment {
  private[this] var mConnectTo: EnumFacing = _
  private[this] var loading = false

  private[this] var fortune = 0
  private[this] var unbreaking = 0
  private[this] var silktouch = false
  private val module: ExpPumpModule = new ExpPumpModule((_: Long) => true, () => this.unbreaking, None)

  override protected def getSymbol: Symbol = BlockExpPump.SYMBOL

  /**
    * Called after enchantment setting.
    */
  override def G_ReInit(): Unit = {
    refreshConnection()
  }

  def getModule = module

  override def onLoad(): Unit = {
    super.onLoad()
    if (loading) {
      refreshConnection()
      loading = false
    }
  }

  private def refreshConnection(): Unit = {
    if (hasWorld && !world.isRemote) {
      val entry = EnumFacing.VALUES
        .map(f => (f, getWorld.getTileEntity(getPos.offset(f))))
        .collectFirst {
          case (f: EnumFacing, t: IAttachable) if t.connect(f.getOpposite, IAttachment.Attachments.EXP_PUMP) => f -> t
        }.getOrElse((null, IAttachable.dummy))

      if (entry._1 != null && entry._2.connectAttachment(entry._1.getOpposite, IAttachment.Attachments.EXP_PUMP, false))
        setConnectTo(entry._1)
      S_sendNowPacket()
    }
  }

  override def setConnectTo(@Nullable connectTo: EnumFacing) {
    this.mConnectTo = connectTo
    if (hasWorld) {
      val state = getWorld.getBlockState(getPos)
      if (!working == state.getValue(BlockPump.CONNECTED)) {
        InvUtils.setNewState(getWorld, getPos, this, state.withProperty(BlockPump.CONNECTED, Boolean.box(working)))
      }
    }
  }

  private def S_sendNowPacket(): Unit = {
    PacketHandler.sendToAround(ExpPumpMessage.create(this), getWorld, getPos)
  }

  def working: Boolean = mConnectTo != null

  def addXp(amount: Int): Unit = {
    module.xp += amount
    if (module.xp > 0 ^ getWorld.getBlockState(getPos).getValue(ADismCBlock.ACTING)) {
      val state = getWorld.getBlockState(getPos).withProperty(ADismCBlock.ACTING, Boolean.box(module.xp > 0))
      InvUtils.setNewState(getWorld, getPos, this, state)
    }
  }

  def getEnergyUse(amount: Int): Long = {
    amount * 10 * APowerTile.MicroJtoMJ / (1 + unbreaking)
  }

  def onActivated(worldIn: World, pos: BlockPos, playerIn: EntityPlayer): Unit = {
    //on Server side
    if (module.xp > 0) {
      val xp = EntityXPOrb.getXPSplit(module.xp)
      val orb = new EntityXPOrb(worldIn, playerIn.posX, playerIn.posY, playerIn.posZ, xp)
      worldIn.spawnEntity(orb)
      addXp(-xp)
    }
  }

  def onBreak(worldIn: World): Unit = {
    if (module.xp > 0) {
      val xpOrb = new EntityXPOrb(worldIn, getPos.getX, getPos.getY, getPos.getZ, module.xp)
      worldIn.spawnEntity(xpOrb)
    }
  }

  /**
    * @return Map (Enchantment id, level)
    */
  override def getEnchantments: util.Map[Integer, Integer] = {
    Map(IEnchantableTile.FortuneID -> fortune,
      IEnchantableTile.UnbreakingID -> unbreaking,
      IEnchantableTile.SilktouchID -> silktouch.compareTo(false))
      .collect(enchantCollector).asJava
  }

  /**
    * @param id    Enchantment id
    * @param value level
    */
  override def setEnchantment(id: Short, value: Short): Unit = {
    id match {
      case IEnchantableTile.FortuneID => fortune = value
      case IEnchantableTile.UnbreakingID => unbreaking = value
      case IEnchantableTile.SilktouchID => silktouch = value > 0
      case _ =>
    }
  }

  override def writeToNBT(compound: NBTTagCompound): NBTTagCompound = {
    compound.setByte("mConnectTo", mConnectTo.map(_.ordinal().toByte).getOrElse(-1))
    compound.setInteger("xpAmount", module.xp)
    compound.setBoolean("silktouch", this.silktouch)
    compound.setByte("fortune", this.fortune.toByte)
    compound.setByte("unbreaking", this.unbreaking.toByte)
    super.writeToNBT(compound)
  }

  override def readFromNBT(compound: NBTTagCompound): Unit = {
    super.readFromNBT(compound)
    val connectID = compound.getByte("mConnectTo")
    mConnectTo = if (connectID < 0) null else EnumFacing.getFront(connectID)
    module.xp = compound.getInteger("xpAmount")
    this.silktouch = compound.getBoolean("silktouch")
    this.fortune = compound.getByte("fortune")
    this.unbreaking = compound.getByte("unbreaking")
    loading = true
  }

  override def getDebugName: String = TranslationKeys.exppump

  /**
    * For internal use only.
    *
    * @return debug info of valid machine.
    */
  override def getDebugMessages: util.List[TextComponentString] = Seq(
    "Connection -> " + mConnectTo,
    "Unbreaking -> " + unbreaking,
    "Fortune -> " + fortune,
    "Silktouch -> " + silktouch,
    "XpAmount -> " + module.xp
  ).map(toComponentString).asJava

  def writeToPacket(message: ExpPumpMessage): ExpPumpMessage = {
    message.pos = getPos
    message.dim = getWorld.provider.getDimension
    message.xpAmount = module.xp
    message.facingOrdinal = Option(mConnectTo).fold(-1)(_.ordinal())
    message
  }

  def onMessage(message: ExpPumpMessage): Unit = {
    mConnectTo = Some(message.facingOrdinal).collect {
      case i if i >= 0 => EnumFacing.getFront(i)
      case _ => null
    }.get
    module.xp = message.xpAmount
  }
}

File: src/main/scala/com/yogpc/qp/machines/exppump/TileExpPump.java
package com.yogpc.qp.machines.exppump;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import com.yogpc.qp.machines.TranslationKeys;
import com.yogpc.qp.machines.base.APacketTile;
import com.yogpc.qp.machines.base.APowerTile;
import com.yogpc.qp.machines.base.IAttachable;
import com.yogpc.qp.machines.base.IAttachment;
import com.yogpc.qp.machines.base.IDebugSender;
import com.yogpc.qp.machines.base.IEnchantableTile;
import com.yogpc.qp.machines.base.IModule;
import com.yogpc.qp.machines.base.QPBlock;
import com.yogpc.qp.utils.Holder;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import net.minecraft.block.BlockState;
import net.minecraft.entity.item.ExperienceOrbEntity;
import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.nbt.CompoundNBT;
import net.minecraft.tileentity.ITickableTileEntity;
import net.minecraft.util.Direction;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.vector.Vector3d;
import net.minecraft.util.text.ITextComponent;
import net.minecraft.world.World;
import org.apache.commons.lang3.tuple.Pair;

import static jp.t2v.lab.syntax.MapStreamSyntax.byValue;
import static jp.t2v.lab.syntax.MapStreamSyntax.values;
import static net.minecraft.state.properties.BlockStateProperties.ENABLED;

public class TileExpPump extends APacketTile implements IEnchantableTile, IDebugSender, IAttachment, ITickableTileEntity {
    @Nullable
    private Direction mConnectTo;
    private ExpPumpModule module = ExpPumpModule.apply(v1 -> true, () -> this.unbreaking);

    private int fortune = 0;
    private int unbreaking = 0;
    private boolean silktouch = false;
    private boolean loading = false;

    public TileExpPump() {
        super(Holder.expPumpTileType());
    }

    @Override
    public void setConnectTo(@Nullable Direction connectTo) {
        mConnectTo = connectTo;
        if (hasWorld()) {
            assert world != null;
            BlockState state = world.getBlockState(getPos());
            if (!working() == state.get(QPBlock.WORKING())) {
                world.setBlockState(pos, state.with(QPBlock.WORKING(), working()));
            }
        }
    }

    @Override
    public IModule getModule() {
        return module;
    }

    @Override
    public String getDebugName() {
        return TranslationKeys.exppump;
    }

    @Override
    public List<? extends ITextComponent> getDebugMessages() {
        return Stream.of(
            "Connection -> " + mConnectTo,
            Stream.of("Unbreaking -> " + unbreaking, "Fortune -> " + fortune, "Silktouch -> " + silktouch).reduce(combiner).get(),
            "XpAmount -> " + module.xp())
            .map(toComponentString)
            .collect(Collectors.toList());
    }

    @Override
    public void G_ReInit() {
        refreshConnection();
    }

    private void refreshConnection() {
        if (hasWorld() && world != null && !world.isRemote) {
            Map.Entry<Direction, IAttachable> entry = Stream.of(Direction.values())
                .map(f -> Pair.of(f, world.getTileEntity(pos.offset(f))))
                .filter(byValue(t -> t instanceof IAttachable))
                .map(values(t -> ((IAttachable) t)))
                .filter(pair ->
                    pair.getValue().connect(pair.getKey().getOpposite(), Attachments.EXP_PUMP)
                )
                .findFirst()
                .orElse(Pair.of(null, IAttachable.dummy));

            if (entry.getKey() == null || entry.getValue().connectAttachment(entry.getKey().getOpposite(), Attachments.EXP_PUMP, false))
                setConnectTo(entry.getKey());
        }
    }

    public void addXp(int amount) {
        if (enabled()) {
            module.xp_$eq(module.xp() + amount);
            assert world != null;
            if (module.xp() > 0 ^ world.getBlockState(pos).get(ENABLED)) {
                BlockState state = world.getBlockState(pos).with(ENABLED, module.xp() > 0);
                world.setBlockState(pos, state);
            }
        }
    }

    public long getEnergyUse(int amount) {
        return amount * 10 * APowerTile.MJToMicroMJ / (1 + unbreaking);
    }

    public void onActivated(World worldIn, BlockPos pos, PlayerEntity playerIn) {
        if (module.xp() > 0) {
            int xp = ExperienceOrbEntity.getXPSplit(module.xp());
            Vector3d vec = playerIn.getPositionVec();
            ExperienceOrbEntity orb = new ExperienceOrbEntity(worldIn, vec.getX(), vec.getY(), vec.getZ(), xp);
            worldIn.addEntity(orb);
            addXp(-xp);
        }
    }

    public void onBreak(World worldIn) {
        if (module.xp() > 0) {
            ExperienceOrbEntity xpOrb = new ExperienceOrbEntity(worldIn, pos.getX(), pos.getY(), pos.getZ(), module.xp());
            worldIn.addEntity(xpOrb);
        }
    }

    @Override
    public void func_230337_a_(BlockState state, CompoundNBT compound) {
        super.func_230337_a_(state, compound);
        int connectID = compound.getByte("mConnectTo");
        mConnectTo = connectID < 0 ? null : Direction.byIndex(connectID);
        module.xp_$eq(compound.getInt("xpAmount"));
        this.silktouch = compound.getBoolean("silktouch");
        this.fortune = compound.getByte("fortune");
        this.unbreaking = compound.getByte("unbreaking");
        loading = true;
    }

    @Override
    public CompoundNBT write(CompoundNBT compound) {
        compound.putByte("mConnectTo", Optional.ofNullable(mConnectTo).map(Direction::ordinal).orElse(-1).byteValue());
        compound.putInt("xpAmount", module.xp());
        compound.putBoolean("silktouch", this.silktouch);
        compound.putByte("fortune", (byte) this.fortune);
        compound.putByte("unbreaking", (byte) this.unbreaking);
        return super.write(compound);
    }

    @Nonnull
    @Override
    public Map<ResourceLocation, Integer> getEnchantments() {
        Map<ResourceLocation, Integer> map = new HashMap<>();
        if (fortune > 0) map.put(FortuneID, fortune);
        if (unbreaking > 0) map.put(UnbreakingID, unbreaking);
        if (silktouch) map.put(SilktouchID, 1);
        return map;
    }

    @Override
    public void setEnchantment(ResourceLocation id, short val) {
        if (id.equals(FortuneID)) {
            fortune = val;
        } else if (id.equals(UnbreakingID)) {
            unbreaking = val;
        } else if (id.equals(SilktouchID)) {
            silktouch = val > 0;
        }
    }

    public boolean working() {
        return mConnectTo != null;
    }

    public int getXpAmount() {
        return module.xp();
    }

    @Override
    public void tick() {
        if (loading) {
            loading = false;
            refreshConnection();
        }
    }
}
